### PR TITLE
CB-841 - checking component statuses before cluster (re)start

### PR DESCRIPTION
**Pull request by attilapalfi92**

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ambari/AmbariClusterModificationService.java [hunk 1]
```diff
@@ -152,16 +174,35 @@ public int startCluster(Stack stack) throws CloudbreakException {
         } else if (isTimeout(ambariHealthCheckResult)) {
             throw new CloudbreakException("Ambari server was not restarted properly.");
         }
-        LOGGER.info("Starting Ambari agents on the hosts.");
         Set<HostMetadata> hostsInCluster = hostMetadataRepository.findHostsInCluster(stack.getCluster().getId());
+        waitForAmbariHosts(stack, ambariClient, hostsInCluster);
+        waitForComponenst(stack, ambariClient, hostsInCluster);
```
Comment (horadla23) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ambari/AmbariClusterModificationService.java]: please fix typo

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ambari/AmbariClusterModificationService.java [hunk 2]
```diff
@@ -89,60 +91,80 @@ public void upscaleCluster(Stack stack, HostGroup hostGroup, Collection<HostMeta
...
         } catch (AmbariConnectionException ignored) {
             LOGGER.debug("Ambari not running on the gateway machine, no need to stop it.");
         }
     }
 
+    private boolean isClusterStopped(AmbariClient ambariClient) {
+        Collection<Map<String, String>> values = ambariClient.getHostComponentsStates().values();
+        for (Map<String, String> value : values) {
+            for (String state : value.values()) {
+                if (!"INSTALLED".equals(state)) {
```
Comment (horadla23) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ambari/AmbariClusterModificationService.java]: maybe it is worth to rewrite to stream filter and check count, isn't it? i hope the size of the components is not that big which can cause performance issues

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/AmbariHostsJoinStatusCheckerTask.java [hunk 3]
```diff
@@ -17,30 +17,36 @@
...
         }
+    }
+
+    private boolean areAllHostsJoined(AmbariHostsCheckerContext ambariHostsCheckerContext) {
+        AmbariClient ambariClient = ambariHostsCheckerContext.getAmbariClient();
+        Map<String, String> hostNamesToStatuses = ambariClient.getHostStatuses();
+        for (HostMetadata host : ambariHostsCheckerContext.getHostsInCluster()) {
+            if (!isHostJoined(host, hostNamesToStatuses)) {
+                LOGGER.info("Host {} is currently not part of the cluster, waiting for it to join.", host.getHostName());
+                return false;
```
Comment (horadla23) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/AmbariHostsJoinStatusCheckerTask.java]: here is also worth to refactor to stream with filter, in this case we can log all of the hosts which isn't joined, not just the first

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ambari/AmbariClusterModificationService.java (hunk 2 above)
Comment (attilapalfi92) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ambari/AmbariClusterModificationService.java]: Well, actually I just extracted this method, and I didn't bother to rewrite it. It is readable, easy to understand and performant.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/AmbariHostsJoinStatusCheckerTask.java (hunk 3 above)
Comment (attilapalfi92) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/AmbariHostsJoinStatusCheckerTask.java]: You are right, this one worth touching. (Also just extracted.)

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ambari/AmbariClusterModificationService.java (hunk 2 above)
Comment (horadla23) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ambari/AmbariClusterModificationService.java]: Ok, we can leave it as-is, if you prefer that

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/AmbariHostsJoinStatusCheckerTask.java (hunk 3 above)
Comment (attilapalfi92) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/AmbariHostsJoinStatusCheckerTask.java]: I am also writing tests for it.

**Pull request merged**